extensions/chathistory.md

```diff
@@ -0,0 +1,158 @@
+---
+title: IRCv3 chathistory extension
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "Evan Magaliff"
+    period: "2017"
+    email: "evan@muffinmedic.net"
+  -
+    name: "Darren Whitlen"
+    period: "2018-2019"
+    email: "darren@kiwiirc.com"
+---
+## Description
+This document describes the format of the `chathistory` extension. This enables clients to request messages that were previously sent if they are still available on the server.
+
+The server as mentioned in this document may refer to either an IRC server or an IRC bouncer.
+
+## Implementation
+The `chathistory` extension uses the [chathistory][batch/chathistory] batch type and introduces a client command, `chathistory`.
+
+To fully support this extension, clients MUST support the [`batch`][batch], [`server-time`][server-time] and [`message-tags`][message-tags] capabilities.
+
+The `chathistory` capability MUST be negotiated. This allows the server and client to act differently when delivering message history on connection.
+
+An ISUPPORT token MUST be sent to the client to state the maximum number of messages a client can request in a single command, represented by an integer. `CHATHISTORY=50`. If `0`, the client SHOULD assume that there is no maximum number of messages.
+
+The `event-playback` capability MAY be negotiated. This allows the client to signal that it is capable of receiving and correctly processing lines that would normally produce a local state change (such as `JOIN` or `MODE`) in its history batches.
+
+### `CHATHISTORY` Command
+`CHATHISTORY` content can be requested by the client by sending the `CHATHISTORY` command to the server. A `batch` MUST be returned by the server. If no content exists to return, an empty batch SHOULD be returned to avoid the client waiting for a reply and to indicate that no content is available.
+
+The `chathistory` command uses the following general syntax structure:
+
+    CHATHISTORY <subcommand> <target> <timestamp | msgid> <limit>
+
+The `target` parameter specifies a single buffer (channel or nickname) from which history SHOULD be retrieved. Any `timestamp` values or parameters MUST be in the format of `YYYY-MM-DDThh:mm:ss.sssZ`.
+
+The special target `*` refers to all direct messages sent to or from the current user, regardless of the other party. This allows the client to retrieve conversations with users it is not yet aware of.
```

Comment:
Could we allow servers to send highlights as part of `*` as well, so that notifications could still work without needing clients to fetch history for all channels?

Reply:
I think this should instead be done by some other mechanism. If I understand correctly, the "*" target denotes targets that the client might not know about. However, the client _should_ know about channels it is in. This may be more applicable to some other subcommand.